usb: device_next: fix CDC NCM datagram pointers processing

jfischer-no opened the pull request:
The implementation should not assume that NDP is placed immediately
after NTH. According to the specification (revision 1.0), NDP can be
anywhere in the transfer buffer. There is also always a "terminating
zero datagram pointer entry", which we should also use to terminate the
datagram processing loop.

Fixes: #81960

jfischer-no committed:
usb: device_next: fix CDC NCM datagram pointers processing

The implementation should not assume that NDP is placed immediately
after NTH. According to the specification (revision 1.0), NDP can be
anywhere in the transfer buffer. There is also always a "terminating
zero datagram pointer entry", which we should also use to terminate the
datagram processing loop.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: make error logging more verbose in CDC NCM

Log errors with LOG_ERR.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: fix block length check in CDC NCM

In the current implementation, the block length in the OUT direciton
should be equal to the transfer length.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: fix CDC NCM GET_NTB_INPUT_SIZE

The device should set the maximum size and maximum number of datagrams
within the IN NTB.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -518,14 +519,15 @@ static int cdc_ncm_acl_out_cb(struct usbd_class_data *const c_data,
 	ret = check_frame(data, buf);
 	if (ret < 0) {
-		LOG_DBG("check frame failed (%d)", ret);
+		LOG_ERR("check frame failed (%d)", ret);
```
What's the real reason behind this? Is this so unexpected condition? Deliberately malformed packets are really common on networks.

I would drop this commit, otherwise the rest seems fine.

jfischer-no replied:
It would not be a mailformed network packet, but a mailformed NTB, so this should be logged as error. We do not check Ethernet packets and do not care about them at all.

jfischer-no committed:
usb: device_next: rework notification handling

We need to handle notifications periodically, for example, when the
interface alternate changes and we also need to track whether the
notification was actually transfered to the host. There is no need to
trigger notification work on iface_start/iface_stop when the USB
interface is not in active configuration. This commit intentianly
changes log level in the notification handling code.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: do not accidentally disable CDC NCM network interface

The Zephyr network interface should not be disabled by the USB device
functions. This sounds confusing, but in the current implementation they
are different ends of the connection.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: rework CDC ECM connected/disconnected notifications

Do not fail in iface_start/iface_stop if sending notification fails,
just log the error and try to send connected notification when alternate
interface is set and network interface is started.

Do not block in notification send function and allow to submit more than
one notification which means that the last valid one will also be
delivered.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: allow CDC ECM to survive alternate toggling

When interface alternate is set back to zero, the implementation still
attempts to submit OUT transfers, but the addressed endpoint is actually
disabled. Change the logic a bit to track when the alternate data
interface is set/reset instead of the class implementation being part of
the active configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: do not stop when OUT transfer has zero length

Align the OUT handling with the CDC NCM and change the logic to track
when the alternate data interface is set/reset instead of the class
implementation being part of the active configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: rework wSequence check in CDC NCM implementation

The consequences of a wSequence mismatch are unspecified and are
primarily for debugging purposes. Our checks are a bit too strict, and
if the header check fails, the sequence is not updated, and the header
check subsequently fails. Rework the code to just log a warning on
mismatch and also reset OUT sequence counter the same way as the IN
sequence counter.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/usb/device_next/class/usbd_cdc_ncm.c:
```diff
@@ -33,7 +33,7 @@ LOG_MODULE_REGISTER(cdc_ncm, CONFIG_USBD_CDC_NCM_LOG_LEVEL);
 enum {
 	CDC_NCM_IFACE_UP,
-	CDC_NCM_CLASS_ENABLED,
+	CDC_NCM_DATA_IFACE_ENABLED,
```
In "usb: device_next: do not stop when OUT transfer has zero length" commit message you cannot really refer to the commit by its SHA, because the SHA will change when the patches enter mainline. The solution would be to split that away from this PR and update the commit message after the underlying commit is in main.

jfischer-no replied:
Yes, I screwed up again. Changed commit message.
